### PR TITLE
Add option to skip null and undefined values in stringify

### LINKED ISSUE #196: `skipNulls` option for queryString.stringify
Hi!

I am now comparing 2 libs for future use on my project: `qs` and `query-string`.
The second one has some advantages like encoding/decoding booleans and numbers which are important for me. But the 1st one has `skipNulls` option which allows to skip `null` and `undefined` values:

```js

const obj1 = {
  a: 1,
  b: '2',
  c: [3, 4, 5],
  d: true,
  e: 'just string',
  f: null,
  g: undefined,
};

// qs
// can skip null/undefined values
const ex2 = qs.stringify(obj1, {
  addQueryPrefix: true,
  arrayFormat: 'comma',
  skipNulls: true,
});
const ex3 = qs.parse(ex2, {
  parseArrays: true,
  comma: true,
  ignoreQueryPrefix: true,
});

// query-string
// can store numbers and booleans
const ex4 = queryString.stringify(obj1, {
  strict: true,
});
const ex5 = queryString.parse(ex4, {
  parseNumbers: true,
  parseBooleans: true,
});

console.log('ex2', ex2);
console.log('ex3', ex3);
console.log('ex4', ex4);
console.log('ex5', ex5);
```

Screenshot from 2019-07-26 10-48-56

### Does it sound reasonable to implement this option?

### PR DESCRIPTION
**Pull request by FredrikAugust**
You can now pass an option called `skipNullAndUndefined` to the options
object in stringify to prevent the generator from generating keys for
those keys.

Fixes #196.

### CODE REVIEW

**Comment by FredrikAugust** [comment]
Appears that `Object.entries` is not supported in older versions of Node.js 6. Will get to this now.

File: index.js [hunk 1]
```diff
@@ -243,6 +243,15 @@ exports.stringify = (object, options) => {
 	}, options);
 
 	const formatter = encoderForArrayFormat(options);
+
+	if (options.skipNullAndUndefined) {
+		for (const key of Object.keys(object)) {
+			if (object[key] === undefined || object[key] === null) {
+				delete object[key];
```
Comment (goodwin64) [inline index.js]: Here you're mutating the passed object.

Comment (goodwin64) [reply index.js]: I suppose `stringify` function has to be pure.

Comment (FredrikAugust) [reply index.js]: Makes sense. Will update it now.

Comment (FredrikAugust) [reply index.js]: Fixed in #d0b21db.

**Comment by sindresorhus** [comment]
@yaodingyd Can you help review? 🙏

File: index.d.ts [hunk 2]
```diff
@@ -100,9 +100,9 @@ The returned object is created with [`Object.create(null)`](https://developer.mo
 
 @param query - The query string to parse.
 */
-export function parse(query: string, options: {parseBooleans: true, parseNumbers: true} & ParseOptions): ParsedQuery<string | boolean | number>;
-export function parse(query: string, options: {parseBooleans: true} & ParseOptions): ParsedQuery<string | boolean>;
-export function parse(query: string, options: {parseNumbers: true} & ParseOptions): ParsedQuery<string | number>;
+export function parse(query: string, options: { parseBooleans: true, parseNumbers: true } & ParseOptions): ParsedQuery<string | boolean | number>;
+export function parse(query: string, options: { parseBooleans: true } & ParseOptions): ParsedQuery<string | boolean>;
+export function parse(query: string, options: { parseNumbers: true } & ParseOptions): ParsedQuery<string | number>;
```
Comment (sindresorhus) [inline index.d.ts]: Don't do unrelated changes. Same in many other places.

File: index.d.ts [hunk 3]
```diff
@@ -192,13 +192,33 @@ export interface StringifyOptions {
 	```
 	*/
 	readonly sort?: ((itemLeft: string, itemRight: string) => number) | false;
+
+	/**
+	Takes a boolean which decides whether or not `undefined` and `null` values should be skipped when stringify-ing.
```
Comment (sindresorhus) [inline index.d.ts]: This could be written better. Try to make it more concise.

File: index.js [hunk 4]
```diff
@@ -243,6 +243,17 @@ exports.stringify = (object, options) => {
 	}, options);
 
 	const formatter = encoderForArrayFormat(options);
+
+	const object = Object.assign({}, input);
+
+	if (options.skipNullAndUndefined) {
+		for (const key of Object.keys(object)) {
+			if (object[key] === undefined || object[key] === null) {
```
Comment (yaodingyd) [inline index.js]: It doesn't work with `arrayFormat`. You need to check if `object[key]` is an array, and iterate through array to check null/undefined.

File: test/stringify.js [hunk 5]
```diff
@@ -193,3 +193,32 @@ test('should disable sorting', t => {
...
+});
+
+test('should ignore both null and undefined when skipNullAndUndefined is set', t => {
+	t.is(queryString.stringify({
+		a: undefined,
+		b: null
+	}, {
+		skipNullAndUndefined: true
+	}), '');
+});
```
Comment (yaodingyd) [inline test/stringify.js]: Add test cases for `arrayFormat`.

Comment (MichaelDeBoey) [reply test/stringify.js]: Maybe also add test cases for when it's set to `false`? 🤔
'Cause I thought this was already the current behavior.

**Comment by sindresorhus** [comment]
Also note https://github.com/sindresorhus/query-string/issues/196#issuecomment-522169947

**Comment by sindresorhus** [comment]
@FredrikAugust Still interested in finishing this?

**Comment by FredrikAugust** [comment]
It would be nice if someone else could take over, don't have much spare time right now to work on this.

**Comment by yaodingyd** [comment]
I can pick this up over the weekend, if it's ok.
